pkg/packetfilter/fake/packetfilter.go

```diff
@@ -0,0 +1,359 @@
+/*
+SPDX-License-Identifier: Apache-2.0
+
+Copyright Contributors to the Submariner project.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package fake
+
+import (
+	"strings"
+	"sync"
+
+	. "github.com/onsi/gomega"
+	"github.com/pkg/errors"
+	"github.com/submariner-io/submariner/pkg/packetfilter"
+	"github.com/submariner-io/submariner/pkg/packetfilter/iptables"
+	"k8s.io/utils/set"
+)
+
+const (
+	remoteCIDRIPSet                    = "SUBMARINER-REMOTECIDRS"
+	localCIDRIPSet                     = "SUBMARINER-LOCALCIDRS"
+	smPostRoutingChain                 = "SUBMARINER-POSTROUTING"
+	smGlobalnetEgressChainForPods      = "SM-GN-EGRESS-PODS"
+	smGlobalnetEgressChainForNamespace = "SM-GN-EGRESS-NS"
+	gnIPSetPrefix                      = "SM-GN-"
+	DriverName                         = "fake"
+)
+
+type PacketFilter struct {
+	mutex                    sync.Mutex
+	chainRules               map[string]set.Set[string]
+	tableChains              map[string]set.Set[string]
+	failOnAppendRuleMatchers []interface{}
+	failOnDeleteRuleMatchers []interface{}
+}
+
+var (
+	tableTypeToStr             = [packetfilter.TableTypeMAX]string{"filter", "mangle", "nat"}
+	iphookChainTypeToStr       = [packetfilter.ChainTypeMAX]string{"filter", "mangle", "nat"}
+	iphookChainTypeToTableType = [packetfilter.ChainTypeMAX]packetfilter.TableType{
+		packetfilter.TableTypeFilter,
+		packetfilter.TableTypeRoute,
+		packetfilter.TableTypeNAT,
+	}
+	chainHookToStr = [packetfilter.ChainHookMAX]string{"PREROUTING", "INPUT", "FORWARD", "OUTPUT", "POSTROUTING"}
+)
+
+func New() *PacketFilter {
+	pf := &PacketFilter{
+		chainRules:  map[string]set.Set[string]{},
+		tableChains: map[string]set.Set[string]{},
+	}
+
+	packetfilter.SetNewDriverFn(func() (packetfilter.Driver, error) {
+		return pf, nil
+	})
+
+	return pf
+}
+
+func (i *PacketFilter) ChainExists(table packetfilter.TableType, chain string) (bool, error) {
+	return i.chainExists(tableTypeToStr[table], chain)
+}
+
+func (i *PacketFilter) AppendUnique(table packetfilter.TableType, chain string, ruleSpec *packetfilter.Rule) error {
+	ruleSpecStr, err := iptables.RuleToRuleSpecFn(ruleSpec)
```

Comment:
We shouldn't have a dependency on iptables. We can simply stringify the `Rule`: `fmt.Sprintf("%+v", ruleSpec)`.